test/Clownface/deleteList.js

```diff
@@ -61,4 +61,12 @@ describe('.deleteList', () => {
     assert.strictEqual(dataset.size, 1)
     assert([...dataset][0].equals(other))
   })
+
+  it('should not hang if enters in a loop ', () => {
+    const dataset = rdf.dataset([
+      rdf.quad(ns.ex.term, ns.ex.p, ns.ex.list)
+    ])
+    const cf = clownface({ term: ns.ex.term, dataset })
+    cf.deleteList(ns.ex.p)
+  })
```

Comment:
I think this test needs a different title and an assertion for when it should succeed

What is the intended outcome? Should `ex:p` triple be removed?

Reply:
I think you are right about the misleading title. 

My first intention was to just avoid clownface hanging when the user tries to delete a non-list or non-well-formed list. (no ending :nil)

But now that you say it, we must define the desired outcome.

I see some branches of how I Imagine this particular operation could work:

- If it's a list
	- Clownface performs the operation (1)
- If it's not a list
	- Clownface throws an error (2)
	- Clownface ignores the operation (3)
- If it's a malformed list
	- Clownface throws an error (4)
	- Clownface ignores the operation (5)

Could be (1), (3) and (4) a reasonable contract?

Reply:
I think I agree, with a small asterisk

"not a list" is technically a "malformed list" IMO, only such that the firs node if malformed.

```
<> 
  ex:notList [ ] ;
  ex:malformedList [
    rdf:first "element" ;
    rdf:rest [] ;
  ] .
```

I would propose a solution where a "malformed list" is ignored when it's already broken on the first node. If the malformation occurs on the `rdf:rest` predicate, then you'd throw for sure

Reply:
So far, so good, 

But on second thought, looking into the isList() and list() method at: https://github.com/zazuko/clownface/blob/a31be80ff34ecfe0b8bdc30249cb2a2cd37164a9/lib/Clownface.js#L140

I think these methods must be reused and their contracts honored. (for example, no multiple values for rdf:rest)